src/ion/merge.rs

```diff
@@ -93,6 +95,17 @@ impl<'a, F: Function> Env<'a, F> {
             }
         }
 
+        // Avoid merging if either side has a fixed-reg def: this can
+        // result in an impossible-to-solve allocation problem if
+        // there is a fixed-reg use in the same reg on the same
+        // instruction.
+        if self.bundles[from.index()].cached_fixed_def()
+            || self.bundles[to.index()].cached_fixed_def()
```

Comment:
If you knew that the fixed def constraints were for the same register, would that change whether or not you needed to bail out here?

Reply:
No, in fact the fuzzbug was exactly this case! The issue is that the way that we do the constraint rewrite adds an anonymous reservation; so this isn't a case of "different fixed def, can't merge" (that is captured by the incompatible-requirement check) but rather "corner case results in impossible problem".

The interesting case is when we have a fixed-use bound to preg P with liverange up to the early point on an inst, a fixed-def bound to preg P (same preg) at the late point on an inst, and that def'd vreg flows into a blockparam arg and over a backedge to the use's vreg. Then we try to merge the two vregs into one bundle (because we try this for every blockparam arg - blockparam def pair). But we rewrote the use to be an "any" use and put an anonymous reservation on preg P at the early-point only. So rejoining the first vreg (the use-side) to the whole bundle and thus requiring it to have the fixed-preg constraint again causes an impossible-to-solve problem.

There may be an even more complex way to rewrite this but I was several levels deep into "introduce fix, get deeper bug" and wanted to bail out of the rabbithole altogether...

Reply:
This check is very conservative and is hurting our compiler quite a bit since it makes heavy use of fixed operands. Removing this check gives a 4% code size reduction (I enabled the checker to ensure the result is still valid).

What are the exact requirements that cause the impossible to resolve constraint? Could this check be made more specific?

Reply:
I don't really have details paged in anymore, beyond the paragraph describing a fuzzing-discovered example I wrote above, nor the cycles at the moment to think through it deeply. You're welcome to remove it and fuzz to (re)discover it, and propose a better fix, of course!